pinecone/core/grpc/index_grpc.py

```diff
@@ -251,11 +245,51 @@ def traceback(self, timeout=None):
 
 class GRPCIndex(GRPCIndexBase):
 
+    """A client for interacting with a Pinecone index via GRPC API."""
+
     @property
     def stub_class(self):
         return VectorServiceStub
 
-    def upsert(self, vectors, async_req=False, **kwargs):
+    def upsert(self,
+               vectors: Union[List[GRPCVector], List[Tuple]],
+               async_req: bool = False,
+               namespace: Optional[str] = None,
+               **kwargs) -> Union[UpsertResponse, PineconeGrpcFuture]:
+        """
+        The upsert operation writes vectors into a namespace.
+        If a new value is upserted for an existing vector id, it will overwrite the previous value.
+
+        Examples:
+            >>> index.upsert([('id1', [1.0, 2.0, 3.0], {'key': 'value'}), ('id2', [1.0, 2.0, 3.0])], namespace='ns1')
+            >>> index.upsert([GRPCVector(id='id1', values=[1.0, 2.0, 3.0], metadata={'key': 'value'}),
+            >>>              GRPCVector(id='id2', values=[1.0, 2.0, 3.0])], async_req=True)
+
+        Args:
+            vectors (Union[List[Vector], List[Tuple]]): A list of vectors to upsert.
+
+                     A vector can be represented by a 1) GRPCVector object or a 2) tuple.
+                     1) if a tuple is used, it must be of the form (id, values, metadata) or (id, values).
+                        where id is a string, vector is a list of floats, and metadata is a dict.
+                        Examples: ('id1', [1.0, 2.0, 3.0], {'key': 'value'}), ('id2', [1.0, 2.0, 3.0])
+
+                    2) if a GRPCVector object is used, a GRPCVector object must be of the form
+                        GRPCVector(id, values, metadata), where metadata is an optional argument of type
+                        Dict[str, Union[str, float, int, bool, List[int], List[float], List[str]]]
+                       Examples: GRPCVector(id='id1', values=[1.0, 2.0, 3.0], metadata={'key': 'value'}),
+                                 GRPCVector(id='id2', values=[1.0, 2.0, 3.0])
+
+                    Note: the dimension of each vector must match the dimension of the index.
+            async_req (bool): If True, the upsert operation will be performed asynchronously.
+                              Defaults to False. [optional]
+            namespace (str): The namespace to write to. If not specified, the default namespace is used. [optional]
+
+        Returns: UpsertResponse, contains the number of vectors upserted
+                 or a PineconeGrpcFuture object if async_req is True.
```

Comment:
Maybe we should add an async example to the `examples` section?  
So that users would know how to poll \ check the async response

Reply:
Also - what happens if the user does `async_req=True` from the Rest API?  
Does he still get a `PineconeGrpcFuture` object? Isn't there some other object coming from OpenAPI?

Reply:
I linked to the documentation about that since it's rarely used